### PR TITLE
Streamline resharding timing + block reward

### PR DESCRIPTION
**Pull request by harmony-ek**
WIP block reward + resharding, for prelim review.  Am working on breaking this me$$ into actually meaningful and independently reviewable PRs.

Gist of this PR:

* The last block of an epoch now contains the resharding assignment for the next epoch, both on beaconchain (where the last block has all committees' resharding assignment) and on regular chains (where the last block has only the local committee's resharding assignment).  In other words, presence of the resharding info on a block marks the end of the current epoch.
* The beaconchain does resharding periodically.
* Regular chains reshard themselves in response to receiving master resharding decision from the beaconchain (currently through a separate shard state message; to be replaced by BC sync once it is fixed).
    * Regular committee leader proposes the end-of-epoch resharding assignment by echoing (copying) it from the master proposal.
    * Regular committee validators compare the proposal against its own copy received independently from beaconchain.
    * Regular committee leader delays its proposal by 5 seconds to give validators a chance to receive the same resharding info that the leader has received.  That is, we allow up 5 seconds from the moment /any/ node of a committee receives the next-epoch sharding info (pessimistically assuming that this first node is the committee leader) and the moment the 2/3 quorum has received the same next-epoch sharding info.
* Block reward code uses this on-chain sharding assignment to figure out which accounts to pay.

### CODE REVIEW

File: consensus/consensus_validator.go [hunk 1]
```diff
@@ -66,6 +67,7 @@ func (consensus *Consensus) processAnnounceMessage(message *msg_pb.Message) {
 
 	copy(consensus.blockHash[:], blockHash[:])
 	consensus.block = block
+	consensus.BlockVerifier = func(*types.Block) error { return nil }
```
Comment (rlan35) [inline consensus/consensus_validator.go]: Is this WIP code?

Comment (harmony-ek) [reply consensus/consensus_validator.go]: Yes, please take a look at the overall direction; the finishing touches are not in.

File: node/node.go [hunk 2]
```diff
@@ -82,6 +88,129 @@ type syncConfig struct {
...
+	// TODO ek – archival
+	if false {
+		cacheConfig = &core.CacheConfig{Disabled: true}
+	}
+	chainConfig := *params.TestChainConfig
+	chainConfig.ChainID.SetUint64(uint64(shardID))
+	bc, err := core.NewBlockChain(
+		db, cacheConfig, &chainConfig, sc.engine, vm.Config{}, nil,
+	)
+    if err != nil {
```
Comment (rlan35) [inline node/node.go]: lint error.

**Comment by rlan35** [comment]
Overall, it's in the right direction. Since it's too much change. I think we should go over this 1-on-1.

File: consensus/consensus.go [hunk 3]
```diff
@@ -265,61 +265,79 @@ func New(host p2p.Host, ShardID uint32, leader p2p.Peer, blsPriKey *bls.SecretKe
...
-		if len(stakeInfos) == 0 {
-			logger.Error("accumulateRewards: node has no stake info",
-				"nodeKey", key.GetHexString())
+	for idx, member := range parentShardState.NodeList {
+		if signed, err := mask.IndexEnabled(idx); err != nil {
+			return ctxerror.New("cannot check for committer bit",
+				"committerIndex", idx,
+			).WithCause(err)
+		} else if !signed {
 			continue
```
Comment (harmony-ek) [inline consensus/consensus.go]:
```suggestion
			continue // TODO ek – pay everyone later
```

File: consensus/consensus_validator.go (hunk 1 above)
Comment (harmony-ek) [reply consensus/consensus_validator.go]: Fix this; it shouldn't be here

File: consensus/consensus_validator.go [hunk 4]
```diff
@@ -91,6 +93,13 @@ func (consensus *Consensus) processAnnounceMessage(message *msg_pb.Message) {
 		utils.GetLogInstance().Warn("Block content is not verified successfully", "error", err)
 		return
 	}
+	if err := consensus.BlockVerifier(&blockObj); err != nil {
```
Comment (harmony-ek) [inline consensus/consensus_validator.go]: TODO – maybe we can do this in commit

File: core/blockchain.go [hunk 5]
```diff
@@ -66,11 +67,12 @@ const (
 	badBlockLimit       = 10
 	triesInMemory       = 128
 	shardCacheLimit     = 2
+	epochCacheLimit     = 10
 
 	// BlocksPerEpoch is the number of blocks in one epoch
 	// currently set to small number for testing
 	// in future, this need to be adjusted dynamically instead of constant
-	BlocksPerEpoch = 10000
+	BlocksPerEpoch = 5
```
Comment (harmony-ek) [inline core/blockchain.go]: FIX THIS

File: core/rawdb/accessors_chain.go [hunk 6]
```diff
@@ -337,6 +340,29 @@ func ReadBlock(db DatabaseReader, hash common.Hash, number uint64) *types.Block
 func WriteBlock(db DatabaseWriter, block *types.Block) {
 	WriteBody(db, block.Hash(), block.NumberU64(), block.Body())
 	WriteHeader(db, block.Header())
+	epoch := block.Header().Epoch
```
Comment (harmony-ek) [inline core/rawdb/accessors_chain.go]: TODO – maybe move into `WriteHeader()`

File: core/rawdb/accessors_chain.go [hunk 7]
```diff
@@ -337,6 +340,29 @@ func ReadBlock(db DatabaseReader, hash common.Hash, number uint64) *types.Block
...
+	if epoch == nil {
+		// backward compatibility
+		return
+	}
+	epoch = new(big.Int).Set(epoch)
```
Comment (harmony-ek) [inline core/rawdb/accessors_chain.go]: TODO – maybe use uint64 for block/epoch #?

File: core/resharding.go [hunk 8]
```diff
@@ -24,7 +27,7 @@ const (
 	// GenesisShardNum is the number of shard at genesis
 	GenesisShardNum = 4
 	// GenesisShardSize is the size of each shard at genesis
-	GenesisShardSize = 50
+	GenesisShardSize = 5
```
Comment (harmony-ek) [inline core/resharding.go]: FIX THIS

File: core/types/block.go [hunk 9]
```diff
@@ -127,6 +120,18 @@ func (h *Header) Size() common.StorageSize {
...
+func (h *Header) Logger(logger log.Logger) log.Logger {
+	return logger.New(
+		"blockHash", h.Hash(),
+		"blockShard", h.ShardID,
+		"blockEpoch", h.Epoch,
+		"blockNumber", h.Number,
+	)
+}
+
+// Log15Ctx returns the contexts
```
Comment (harmony-ek) [inline core/types/block.go]: Remove this

File: node/node.go [hunk 10]
```diff
@@ -242,39 +269,24 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, db ethdb.Database, is
 		node.SelfPeer = host.GetSelfPeer()
 	}
 
+	// Create test keys.  Genesis will later need this.
+	node.TestBankKeys, err = CreateTestBankKeys(FakeAddressNumber)
```
Comment (harmony-ek) [inline node/node.go]: This is txgen-specific; remove this (add todo)

File: node/node.go [hunk 11]
```diff
@@ -242,39 +269,24 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, db ethdb.Database, is
...
+	if err != nil {
+		utils.GetLogInstance().Crit("Error while creating test keys",
+			"error", err)
+	}
+
+	// TODO ek – make db directory configurable
```
Comment (harmony-ek) [inline node/node.go]: Remove this; it's done

File: node/node_newblock.go [hunk 12]
```diff
@@ -50,25 +57,30 @@ func (node *Node) WaitForConsensusReady(readySignal chan struct{}, stopChan chan
 					threshold = FirstTimeThreshold
 					firstTime = false
 				}
-				if len(node.pendingTransactions) >= threshold {
-					utils.GetLogInstance().Debug("PROPOSING NEW BLOCK ------------------------------------------------", "blockNum", node.blockchain.CurrentBlock().NumberU64()+1, "threshold", threshold, "pendingTransactions", len(node.pendingTransactions))
+				if len(node.pendingTransactions) >= threshold || !time.Now().Before(deadline) {
+					deadline = time.Now().Add(10 * time.Second)
```
Comment (harmony-ek) [inline node/node_newblock.go]: Make this a named constant

File: node/node_newblock.go [hunk 13]
```diff
@@ -50,25 +57,30 @@ func (node *Node) WaitForConsensusReady(readySignal chan struct{}, stopChan chan
...
+					utils.GetLogInstance().Debug("PROPOSING NEW BLOCK ------------------------------------------------", "blockNum", node.Blockchain().CurrentBlock().NumberU64()+1, "threshold", threshold, "pendingTransactions", len(node.pendingTransactions))
```
Comment (harmony-ek) [inline node/node_newblock.go]: Go through this with @chaosma

File: node/node_newblock.go [hunk 14]
```diff
@@ -85,11 +97,56 @@ func (node *Node) WaitForConsensusReady(readySignal chan struct{}, stopChan chan
...
+	shardState, err := core.CalculateNewShardState(
+		node.Blockchain(), nextEpoch, &node.CurrentStakes)
+	if err != nil {
+		return err
+	}
+	block.AddShardState(shardState)
+	return nil
+}
+
+func (node *Node) proposeLocalShardState(block *types.Block) {
```
Comment (harmony-ek) [inline node/node_newblock.go]: Retry instead of delaying

File: consensus/consensus_validator.go (hunk 1 above)
Comment (chaosma) [reply consensus/consensus_validator.go]: this is the equivalent of onAnnounce

File: node/node_newblock.go (hunk 12 above)
Comment (chaosma) [reply node/node_newblock.go]: change similarly to WaitForConsensusReadyv2

**Comment by harmony-ek** [comment]
Jenkins build/test failed after rebasing onto @chaosma's view change.  Since he is working on the view change code, I'd suggest unify the debugging effort on that front.  Request to override Jenkins failure (Travis passed).  Cc: @LeoHChen

**Comment by LeoHChen** [comment]
Jenkins, test this please.

**Comment by LeoHChen** [comment]
@harmony-ek does it mean the txgen won't work anymore after this PR?

**Comment by LeoHChen** [comment]
still conflicting ...

**Pull request merged**